js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -0,0 +1,591 @@
+        ? O
+        : typeof options.getImplicitPkField extends Function
+        ? O
+        : { id: NumberField } & O;
+    type PkFieldType = typeof options.pkFieldName extends string
+        ? string
+        : typeof options.pkFieldName extends string[]
+        ? string[]
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldCompound<O>
+        ? string[]
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldSingle<O>
+        ? string
+        : 'id';
+
+    type PkValueType = typeof options.pkFieldName extends string
+        ? SinglePrimaryKey
+        : typeof options.pkFieldName extends string[]
+        ? CompoundPrimaryKey
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldCompound<O>
+        ? CompoundPrimaryKey
+        : typeof options.getImplicitPkField extends GetImplicitPkFieldSingle<O>
+        ? SinglePrimaryKey
+        : SinglePrimaryKey;
+
+    // This is the constructor function for the created class
+    function _Base<T extends FinalFields = FinalFields>(
+        data: FieldDataMappingRaw<T>
+    ): FieldDataMapping<T> {
+        const pkFieldNames = this._model.pkFieldNames;
+        const missing = pkFieldNames.filter(name => !(name in data));
+        const empty = pkFieldNames.filter(name => name in data && data[name] == null);
+        const errors: string[] = [];
+        if (empty.length > 0) {
+            errors.push(
+                `Primary key(s) '${empty.join("', '")}' was provided but was null or undefined`
+            );
+        }
+        if (missing.length > 0) {
+            errors.push(`Missing value(s) for primary key(s) '${missing.join("', '")}'`);
+        }
+
+        if (errors.length) {
+            throw new Error(errors.join(', '));
+        }
+
+        const assignedData: Record<string, any> = {};
+        const assignedFields: string[] = [];
+        const fields = this._model.fields;
+        for (const [key, value] of Object.entries(data)) {
+            const field = fields[key];
+            if (field) {
+                assignedData[key] = field.normalize(value);
+                assignedFields.push(key);
+            } else {
+                // TODO: Should extra keys in data be a warning or ignored?
+                console.warn(
+                    `Received value for key ${key}. No such field exists on ${this._model.name}`
+                );
+            }
+        }
+
+        this._assignedFields = assignedFields;
+        this._assignedFields.sort();
+        this._data = freezeObject(assignedData);
+
+        // Add getters for fields along with some traps for bad usage
+        // - If attempts to set a field throw an error in dev or log warning otherwise
+        // - If attempts to get a field that does exist but wasn't passed in `data` then
+        //   throw an error in dev or log a warning otherwise.
+        for (const fieldName of Object.keys(fields)) {
```

Comment:
Can you double check whether this needs to be on each instance instantiation or whether the definition can go once on the class

Reply:
I've moved this to the prototype